src/hardware/sblaster.cpp

```diff
@@ -294,6 +308,129 @@ static void InitializeSpeakerState()
 	}
 }
 
+static void configure_filters(Section_prop* config)
+{
+	auto get_filter_params = [&](const char * conf) {
+		static const std::map<SB_TYPES, FilterType> sb_type_to_filter_type_map = {
+				{SBT_NONE, FilterType::None},
+				{SBT_1, FilterType::SB1},
+				{SBT_2, FilterType::SB2},
+				{SBT_PRO1, FilterType::SBPro1},
+				{SBT_PRO2, FilterType::SBPro2},
+				{SBT_16, FilterType::SB16},
+				{SBT_GB, FilterType::None},
+		};
+
+		static const std::map<std::string, FilterType> filter_map = {
+				{"none", FilterType::None},
+				{"sb1", FilterType::SB1},
+				{"sb2", FilterType::SB2},
+				{"sbpro1", FilterType::SBPro1},
+				{"sbpro2", FilterType::SBPro2},
+				{"sb16", FilterType::SB16},
+		};
+		assert(conf);
+		const auto filter_prefs = split(config->Get_string(conf));
+		const auto filter = filter_prefs.empty() ? "auto" : filter_prefs[0];
+
+		FilterType filter_type = FilterType::None;
+		if (filter == "auto") {
+			auto it = sb_type_to_filter_type_map.find(sb.type);
+			if (it != sb_type_to_filter_type_map.end())
+				filter_type = it->second;
+		} else {
+			auto it = filter_map.find(filter);
+			if (it != filter_map.end())
+				filter_type = it->second;
+		}
+
+		FilterState filter_state = FilterState::Off;
+		if (filter_type != FilterType::None) {
+			if (filter_prefs.size() > 1 && filter_prefs[1] == "always_on") {
+				filter_state = FilterState::ForcedOn;
+			} else {
+				filter_state = FilterState::On;
+			}
+		}
+		return std::make_tuple(filter_type, filter_state);
+	};
+	using std::tie;
+	tie(sb.sb_filter_type, sb.sb_filter_state) = get_filter_params("sb_filter");
+	tie(sb.opl_filter_type, std::ignore) = get_filter_params("opl_filter");
+}
+
+static void set_sb_filter()
+{
+	auto set_filter = [](const uint8_t order, const uint16_t cutoff_freq) {
+		sb.chan->ConfigureLowPassFilter(order, cutoff_freq);
+		sb.chan->SetLowPassFilter(sb.sb_filter_state);
+	};
+
+	auto disable_filter = [] { sb.chan->SetLowPassFilter(FilterState::Off); };
+
+	auto enable_zoh_upsampler = [] {
+		constexpr auto dac_rate = 45454;
+		sb.chan->ConfigureZeroOrderHoldUpsampler(dac_rate);
+		sb.chan->EnableZeroOrderHoldUpsampler();
+	};
+
+	switch (sb.sb_filter_type) {
+	case FilterType::None:
+		disable_filter();
+		enable_zoh_upsampler();
+		break;
```

Comment:
If the filter's disabled, should the upsampler also be too? (similar to the SB16 case?)

Reply:
No, the upsampling should definitely happen in the filter disabled mode.

All SB boards upsample to 45454 Hz internally using zero-order-hold, then the pre-SB16 models have a software toggleable fixed lowpass filter at the output. Naturally, this simplistic upsampling results in _tons_ of aliasing, but that's the iconic metallic/grungy/trashy early Sound Blaster sound for you. If the filter is disabled in software, the DAC still outputs the crappily upsampled 45454 Hz stream, without any antialiasing filter whatsoever! That's what we're emulating here, and quite faithfully.

Whereas in the SB16 case we disable the lowpass filter (from the `iir` lib) and the crappy zero-order-hold upsampler, and just let the brickwall lowpass filter built-in into the speex resampler do its thing (which cannot be turned off, by the way). So for example, if the SB mixer rate is 10kHz, then speex upsamples this to our host rate of say 48kHz and applies a brickwall filter at 10 / 2 = 5kHz — perfect! That's virtually indistinguishable from what the SB16 is doing, I've even included screenshots of this behaviour in my ticket that introduced Speex https://github.com/dosbox-staging/dosbox-staging/pull/1687

So effectively by only letting Speex do its thing, we get a very close emulation of the SB16 sound. (Yes, it's not 100% identical, because in reality the SB16 is most likely still doing the zero-order-hold upsampling first, then it applies the variable brickwall filter based on the SB rate, and that's not quite as good quality (it's less steep) than the one Speex uses (you can actually see the difference in steepness on a spectrum analyser when you compare it against real hardware recordings). But to my ears the results are virtually indistinguishable from the real thing if you're just listening to it and are not staring at charts, so I consider this to be a perfectly good emulation. Technically we could always upsample to 45454 Hz, even for the SB16, the apply a bit less steep brickwall using the `iir` lib or whatever, but I find this to be a bit pointless, honestly...)

Reply:
Thanks for the explanation -and makes perfect sense.